cpp/include/raft/neighbors/detail/ivf_pq_compute_similarity-inl.cuh

```diff
@@ -0,0 +1,845 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/distance/distance_types.hpp>  // raft::distance::DistanceType
+#include <raft/matrix/detail/select_warpsort.cuh>  // matrix::detail::select::warpsort::warp_sort_distributed
+#include <raft/neighbors/detail/ivf_pq_dummy_block_sort.cuh>  // dummy_block_sort_t
+#include <raft/neighbors/ivf_pq_types.hpp>                    // codebook_gen
+#include <raft/util/cuda_rt_essentials.hpp>                   // RAFT_CUDA_TRY
+#include <raft/util/device_atomics.cuh>                       // raft::atomicMin
+#include <raft/util/pow2_utils.cuh>                           // raft::Pow2
+#include <raft/util/vectorized.cuh>                           // raft::TxN_t
+#include <rmm/cuda_stream_view.hpp>                           // rmm::cuda_stream_view
+
+namespace raft::neighbors::ivf_pq::detail {
+
+/**
+ * Maximum value of k for the fused calculate & select in ivfpq.
+ *
+ * If runtime value of k is larger than this, the main search operation
+ * is split into two kernels (per batch, first calculate distance, then select top-k).
+ */
+static constexpr int kMaxCapacity = 128;
+static_assert((kMaxCapacity >= 32) && !(kMaxCapacity & (kMaxCapacity - 1)),
+              "kMaxCapacity must be a power of two, not smaller than the WarpSize.");
+
+// using weak attribute here, because it may be compiled multiple times.
```

Comment:
Just curious: what are the reason for using `__attribute__((weak))` as opposed to `inline`?

Reply:
Inline implies that:
1. the function may be defined in multiple translation units
2. the function symbol may be elided from the symbol table of the object file

In this case, we want only behavior 1, not 2. The `is_local_topk_feasible` function is also declared in `-ext` so other translation units may depend on its existence for linking. 

The reason we cannot use `RAFT_INLINE_CONDITIONAL` is that we have multiple translation units that define `is_local_topk_feasible` (one for each instance of `ivf_pq_compute_similarity`). We would run into multiple definition errors during linking if we did not specify the symbol as weak.

Reply:
Thanks for the explanation!